src/args.c

```diff
@@ -77,6 +78,29 @@ int ashuffle_init(struct ashuffle_options * opts) {
     return 0;
 }
 
+/* "safe" string to unsigned conversion
+ * unlike strtoul, considers partial matches e.g. "42foo" or "" as errors.
+ * returns UINT_MAX and sets errno on error.
+ */
+unsigned strtou(char * str) {
```

Comment:
Why not just return a struct here? Since we only use this for parsing args the small additional overhead is really not a big deal.

Otherwise, I think users can probably live with only being able to keep UINT_MAX-1 songs in the queue buffer :)

Reply:
hmm, I don't really like returning structs in C;  just tried and gcc does not complain about it but I don't see any such function here and would rather not be first :p
I guess we could use errno as you used to, and just clear it in case of no error in strtou, but given that's an unsafe pattern for strtoul I don't really want to give people the impression this is a good idea in general.


Personal feelings asides I also think this makes it slightly closer to strtoul which reserves ULONG_MAX + setting errno to ERANGE in case the value would overflow.
Since we're clipping the result ourselves here it would be safe to use UINT_MAX... I'm just a bit reluctant and you weren't insistent enough in that suggestion for me to act on it; ask again if you think it's better ;)

Reply:
> I don't really like returning structs in C

I never understood this about C programmers :smile:. In the right situation, seems better than using special values.

Either way, if it’s closer to the strtoul implementation that’s good enough for me.

Reply:
It's really weird, I have no problem doing that in most other languages, and even agree it's usually better :) (and it's pushed to the extreme with result/option types in stuff like rust and ocaml, which I actually do appreciate as well)

I'm actually not sure where this comes from, I guess it's a cycle of never seeing this anywhere so not wanting to use it... If the struct is big there is overhead but for four bytes it's really not making any difference, so it's really just style at this point.

Thanks for merging anyway, back and forth is fine with me :)